sematic/db/migrations/20230706150252_pre_alembic_parity.py

```diff
@@ -0,0 +1,339 @@
+            )
+            conn.execute("INSERT INTO metric_values_new SELECT * FROM metric_values;")
+            conn.execute("DROP TABLE metric_values;")
+            conn.execute(
+                "CREATE INDEX metric_values_id_time_idx ON metric_values_new (metric_id, metric_time DESC);"
+            )
+            conn.execute(
+                "CREATE INDEX metric_values_time_idx ON metric_values_new (metric_time DESC);"
+            )
+            conn.execute("ALTER TABLE metric_values_new RENAME TO metric_values;")
+    else:
+        with db().get_engine().begin() as conn:
+            conn.execute(
+                """
+                UPDATE runs SET function_path = 'UNKNOWN.UNKNOWN' WHERE function_path IS NULL;
+                UPDATE runs SET tags = '[]' WHERE tags IS NULL;
+                UPDATE runs SET source_code = 'source code unavailable' WHERE source_code IS NULL;
+                UPDATE jobs SET created_at = to_timestamp(0) WHERE created_at IS NULL;
+                UPDATE jobs SET updated_at = to_timestamp(0) WHERE updated_at IS NULL;
+                UPDATE notes SET created_at = to_timestamp(0) WHERE created_at IS NULL;
+                UPDATE notes SET updated_at = to_timestamp(0) WHERE updated_at IS NULL;
+
+                ALTER TABLE artifacts ALTER COLUMN type_serialization SET NOT NULL;
+                ALTER TABLE jobs ALTER COLUMN created_at SET NOT NULL;
+                ALTER TABLE jobs ALTER COLUMN updated_at SET NOT NULL;
+                ALTER TABLE notes ALTER COLUMN created_at SET NOT NULL;
+                ALTER TABLE notes ALTER COLUMN updated_at SET NOT NULL;
+                ALTER TABLE runs ALTER COLUMN function_path SET NOT NULL;
+                ALTER TABLE runs ALTER COLUMN root_id SET NOT NULL;
+                ALTER TABLE runs ALTER COLUMN tags SET NOT NULL;
+                ALTER TABLE runs ALTER COLUMN source_code SET NOT NULL;
+
+                ALTER TABLE edges ADD CONSTRAINT edges_destination_run_id_fkey FOREIGN KEY (destination_run_id) REFERENCES runs(id);
+                ALTER TABLE edges ADD CONSTRAINT edges_source_run_id_fkey FOREIGN KEY (source_run_id) REFERENCES runs(id);
+
+                ALTER TABLE metric_values ADD CONSTRAINT metric_values_metric_id_fkey FOREIGN KEY (metric_id) REFERENCES metric_labels(metric_id);
+                ALTER TABLE runs ADD CONSTRAINT runs_root_id_fkey FOREIGN KEY (root_id) REFERENCES runs(id);
+
+                ALTER TABLE runs DROP COLUMN exception;
+                ALTER TABLE runs DROP COLUMN external_jobs_json;
+
+                ALTER TABLE resolutions DROP COLUMN external_jobs_json;
+
+                ALTER INDEX jobs_run_id RENAME TO ix_jobs_run_id;
+                ALTER INDEX runs_cache_key_index RENAME TO ix_runs_cache_key;
+                ALTER INDEX runs_calculator_path RENAME TO ix_runs_function_path;
+                """
+            )
+
+
+def down():
+    if db().get_engine().url.drivername == "sqlite":
+        pass
```

Comment:
In `up` for sqlite please make a backup copy of the sqlite db file (e.g. `db.sqlite3` -> `db.sqlite3_0.31.2.bck`). The file location is defined [here](https://github.com/sematic-ai/sematic/blob/2130b05f6187fee38d138e5c45f75e0b2bf09342/sematic/config/config.py#L211-L212), and you can use [this incantation](https://github.com/sematic-ai/sematic/blob/2130b05f6187fee38d138e5c45f75e0b2bf09342/sematic/db/db.py#L69) to programmatically source it.

In `down` for sqlite, please try to reinstate the backup, if it exists and is not empty (you can make another copy of `db.sqlite3` for extra safety).

If you get I/O errors in any of the operations, please log an error and continue executing, hoping for the best.

Reply:
You can just take [this code](https://github.com/sematic-ai/sematic/pull/917/files#diff-c4468b27d2c82cc91efd170976d17396116f70a0456017d026a25c5dcb385cd4R16-R69). I'll re-rebase and deduplicate in my PR.

Reply:
thanks!  copied the code into my migration, please rebase and dedupe in your changes whenever you're ready.